oak_runtime/src/node/http/tests.rs

```diff
@@ -244,12 +265,30 @@ async fn send_request(
     let client: hyper::client::Client<_, hyper::Body> =
         hyper::client::Client::builder().build(https);
 
-    let request = hyper::Request::builder()
-        .method(http::Method::GET)
-        .uri(uri)
-        .header(oak_abi::OAK_LABEL_HTTP_PROTOBUF_KEY, label_bytes)
-        .body(hyper::Body::empty())
-        .unwrap();
-
-    client.request(request).await
+    // The client thread may start sending the requests before the server is up. In this case, the
+    // request will be rejected with a "ConnectError". To make the tests are stable, we need to
+    // retry sending the requests until the server is up. To distinguish between these cases and
+    // actual errors (e.g., errors due to invalid TLS certificates), we need to check the cause of
+    // the error.
+    loop {
+        let request = hyper::Request::builder()
```

Comment:
This logic seems very coupled with the actual testing logic now, and also since requests may not be idempotent, it seems a bit risky to just keep retrying. 

Have you considered a different approach, in which *before* sending the actual request, the client performs some other sort of health check? that may be just sending a [`CONNECT`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) request until it succeeds, and then send the actual request for test after that. Alternatively expose another endpoint (e.g. `/status` or `/health`, not sure what's the convention) that returns a value when the server is up. Presumably we will need that for proper cloud deployments too anyways, right?

Reply:
I don't see how `CONNECT` would help. We would still need to retry on that. In general, retrying should be part of any robust interaction with the server, so in my opinion the coupling is normal. 

What do you mean by idempotent "request"? Please note that this is not the general HTTP server pseudo node logic Tis is just a specific test. The GET method that is called here is idempotent (is not it? It does not do anything really!). Moreover, the tests do not discard the actual responses they get from the server. So, it would not be a problem even if the GET method used in the tests was not idempotent. 

Exposing another endpoint would still require the same retrying logic, but for the health check endpoint. So, again I don't see how that could have any impact apart from making the tests unnecessarily complicated and diverted from the important scenario that needs to be tested. 

Also, please note that the health check endpoints should be added by each Oak node connected to an HTTP server pseudo node separately, and what we have here is just a test. We could perhaps modify HTTP server pseudo node to add those endpoints by default or if configured to do so. That would address the cloud deployment need for it, but is out of the scope of this test.

Reply:
Thanks for the detailed reply, I learned a few things from it.

> This does not generally guarantee a robust behaviour. Even if the health check request is successful, there is no guarantee that the server will remain up and running for the next request. In a robust application each request must be retried regardless of the status from the health check or previous requests to the server. The scenario you have described is not realistic, and I think it is better to simulate a realistic scenario in the test.

Note that we don't expect flakiness during test, so retrying requests just in case they fail because the server went down again seems like something we would want to fail on, rather than have retrying logic around. I still think what I described is realistic for a test: wait until the server is up (before performing any request), then expect that every subsequent request gets served correctly at the first try.

> This is not true. The retry loop does not retry the entire test logic, it only retries sending the request, which as described above is what is needed to be done in reality. I have now moved the creation of the client out of send_request. So it should be clearer that the retry loop only retries sending the request.

Does this imply that for any HTTP request done by a real client, it should always be retried in a loop because it may fail?

Reply:
> Does this imply that for any HTTP request done by a real client, it should always be retried in a loop because it may fail?

Yes. It should be fixed in the rust client in `http_server` too, but in a different PR. Hyper's `client::request` already retries sending the request, but retries on other types of errors.

Reply:
ah nice, I didn't know that!